src/Razor/src/Microsoft.AspNetCore.Razor.Language/Components/ComponentReferenceCaptureLoweringPass.cs

```diff
@@ -33,32 +35,106 @@ protected override void ExecuteCore(RazorCodeDocument codeDocument, DocumentInte
+                    // already be a diagnostic for empty values
+                    var identifier = DetermineIdentifierToken(node);
+                    if (identifier == null)
+                    {
+                        continue;
+                    }
+
+                    var rewritten = RewriteUsage(reference.Parent, identifier);
+                    reference.Replace(rewritten);
+
+                    // Now we need to check if the field generation has been suppressed.
+                    //
+                    // You have to suppress field generation for generic types because we don't know the
+                    // type name to create the field.
+                    var generateField = ShouldGenerateField(reference.Parent);
+
+                    // Insert the field with other fields, near the top of the class.
+                    if (generateField)
+                    {
+                        var position = 0;
+                        while (position < @class.Children.Count && @class.Children[i] is FieldDeclarationIntermediateNode)
+                        {
+                            i++;
+                        }
+
+                        @class.Children.Insert(position, CreateField(rewritten.FieldTypeName, identifier));
+                    }
                 }
             }
         }
 
-        private IntermediateNode RewriteUsage(ClassDeclarationIntermediateNode classNode, IntermediateNode parent, TagHelperDirectiveAttributeIntermediateNode node)
+        private ReferenceCaptureIntermediateNode RewriteUsage(IntermediateNode parent, IntermediateToken identifier)
         {
-            // If we can't get a nonempty attribute name, do nothing because there will
-            // already be a diagnostic for empty values
-            var identifierToken = DetermineIdentifierToken(node);
-            if (identifierToken == null)
-            {
-                return node;
-            }
-
             // Determine whether this is an element capture or a component capture, and
             // if applicable the type name that will appear in the resulting capture code
             var componentTagHelper = (parent as ComponentIntermediateNode)?.Component;
             if (componentTagHelper != null)
             {
-                return new ReferenceCaptureIntermediateNode(identifierToken, componentTagHelper.GetTypeName());
+                return new ReferenceCaptureIntermediateNode(identifier, componentTagHelper.GetTypeName());
             }
             else
             {
-                return new ReferenceCaptureIntermediateNode(identifierToken);
+                return new ReferenceCaptureIntermediateNode(identifier);
+            }
+        }
+
+        private bool ShouldGenerateField(IntermediateNode parent)
+        {
+            var parameters = parent.FindDescendantNodes<TagHelperDirectiveAttributeParameterIntermediateNode>();
```

Comment:
We should expand on `DocumentIntermediateNodeExtensions` to include a `FindDescendantReferences` that doesn't require the `DocumentIntermediateNode`. That way we could use the intermediate node reference type below.

Reply:
I tried this actually as my first attempt. The problem is that the `FindDescendantReferences` extension defined on `DocumentIntermediateNode` then becomes ambigious. We messed up by not having this from the beginning.

So I decided to punt on it because the workaround is easy.

One solution would be to define these as *non-extension* methods, then they would take priority. I don't really want to mess with it in this PR because it's not that important right now IMO.